### PR TITLE
Track which Widget objects were created by the local project.

### PR DESCRIPTION
**Pull request by jacob314**

### CODE REVIEW

**Review by pq — APPROVED** [review]
👍

**Comment by Hixie** [comment]
Tests?

**Comment by jacob314** [comment]
Modified `flutter test` to support the `--track-widget-creation`
and added tests to `widget_inspector_test.dart` that use this functionality to test
creation location tracking including tracking whether widgets were created by the current project.
I can move the flutter_tools change to a separate CL if desired. Keeping it here shows how the small new flutter_tools functionality is used.

Note that there are still about 17 tests that still fail when the --preview-dart-2 flag is set. I don't know how long the remaining failing tests are expected to be failing for. If they can't be fixed soon the widget_inspector_test could be run along side the benchmark script with --preview-dart-2 and --track-widget-creation. Long term I think all flutter tests should be run with --track-widget-creation set as well as with it not set to ensure the --track-widget-creation flag is not introducing differences in behavior.

Examples of how to run and how not to run these tests today:
```
jacobr-macbookpro2:flutter jacobr$ flutter test  --track-widget-creation test/widgets/widget_inspector_test.dart
Building flutter tool...
Running "flutter packages get" in flutter...                 0.6s
--track-widget-creation is valid only when --preview-dart-2 is specified.

Run 'flutter -h' (or 'flutter <command> -h') for available flutter commands and options.
jacobr-macbookpro2:flutter jacobr$ flutter test --preview-dart-2 --track-widget-creation test/widgets/widget_inspector_test.dart
00:04 +0: loading /Users/jacobr/git/flutter/2/flutter/packages/flutter/test/widgets/widget_inspector_test.dart
Finished compilation of /var/folders/n5/d160x2xj2095jng193ptjjl00025bq/T/dart_test_listenerU5e6cB/listener.dart into /var/folders/n5/d160x2xj2095jng193ptjjl00025bq/T/dart_test_listenerU5e6cB/listener.dart.dill
00:06 +16: All tests passed!                                                   
jacobr-macbookpro2:flutter jacobr$ flutter test test/widgets/widget_inspector_test.dart
00:02 +14: - WidgetInspectorService creationLocation                           
Skipping WidgetInspectorService creationLocation test because --track-widget-creation flag not set.
00:02 +15: - WidgetInspectorService setPubRootDirectories                      
WidgetInspectorService setPubRootDirectories test skipped because --track-widget-creation flag was not set.
00:02 +16: All tests passed!                                                   
jacobr-macbookpro2:flutter jacobr$
```

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 1]
```diff
@@ -23,54 +23,71 @@ class TestCommand extends FlutterCommand {
...
+    argParser.addFlag(
+      'start-paused',
+      defaultsTo: false,
+      negatable: false,
+      help: 'Start in a paused mode and wait for a debugger to connect.\n'
+            'You must specify a single test file to run, explicitly.\n'
+            'Instructions for connecting with a debugger and printed to the\n'
+            'console once the test has started.',
     );
-    argParser.addFlag('coverage',
```
Comment (jacob314) [inline packages/flutter_tools/lib/src/commands/test.dart]: the only real change is adding the  'track-widget-creation' flag
the other changes are just switching the inconsistent idents to match the flutter style guide.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 2]
```diff
@@ -100,7 +102,13 @@ class _FlutterPlatform extends PlatformPlugin {
     this.explicitObservatoryPort,
     this.host,
     this.previewDart2,
-  }) : assert(shellPath != null) {
+    @required this.trackWidgetCreation,
+  }) : assert(shellPath != null),
+       compiler = new ResidentCompiler(
```
Comment (jacob314) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: the resident compiler initialization code is unchanged except for passing in the trackWidgetCreation flag.

**Comment by jacob314** [comment]
@aam can you review the changes under  packages/flutter_tools?

**Review by aam — APPROVED** [review]

File: packages/flutter/test/widgets/widget_inspector_test.dart [hunk 3]
```diff
@@ -498,4 +498,157 @@ void main() {
       expect(service.toObject(propertyJson['objectId']), const isInstanceOf<DiagnosticsNode>());
     }
   });
+
+  testWidgets('WidgetInspectorService creationLocation', (WidgetTester tester) async {
+    final WidgetInspectorService service = WidgetInspectorService.instance;
+    if (!service.isWidgetCreationTracked()) {
+      print('Skipping WidgetInspectorService creationLocation test '
```
Comment (Hixie) [inline packages/flutter/test/widgets/widget_inspector_test.dart]: use the `skip` argument to `testWidgets` to do this.

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter_tools/lib/src/commands/test.dart]: thanks

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 4]
```diff
@@ -23,54 +23,71 @@ class TestCommand extends FlutterCommand {
...
+    argParser.addFlag(
+      'preview-dart-2',
+      hide: !verboseHelp,
+      help: 'Preview Dart 2.0 functionality.',
+    );
+    argParser.addFlag(
+      'track-widget-creation',
+      negatable: false,
+      hide: !verboseHelp,
+      help: 'Track widget creation locations.',
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/test.dart]: maybe add some comment like "This enables testing of features such as the widget inspector."

File: packages/flutter_tools/lib/src/test/runner.dart [hunk 5]
```diff
@@ -29,8 +30,14 @@ Future<int> runTests(
     bool ipv6: false,
     bool machine: false,
     bool previewDart2: false,
+    bool trackWidgetCreation: false,
     TestWatcher watcher,
     }) async {
+  if (trackWidgetCreation && !previewDart2) {
+    throw new UsageException(
+        '--track-widget-creation is valid only when --preview-dart-2 is specified.', null);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/test/runner.dart]: you can make this a one-liner if you prefer. Or alternatively:
```dart
  throw new UsageException(
    '--track-widget-creation is valid only when --preview-dart-2 is specified.',
     null,
  );
￼```

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/test/widgets/widget_inspector_test.dart (hunk 3 above)
Comment (jacob314) [reply packages/flutter/test/widgets/widget_inspector_test.dart]: nice! done.

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 4 above)
Comment (jacob314) [reply packages/flutter_tools/lib/src/commands/test.dart]: done

File: packages/flutter_tools/lib/src/test/runner.dart (hunk 5 above)
Comment (jacob314) [reply packages/flutter_tools/lib/src/test/runner.dart]: done

**Pull request merged**